ci: run integration test on github actions

subhamkrai opened the pull request:
**Description of your changes:**

as we are planning to move toward Github actions
for integration test. This commit runs ceph 
suites using Github action.

Signed-off-by: subhamkrai <srai@redhat.com>

<!-- Please take a look at our [Contributing](https://rook.io/docs/rook/master/development-flow.html)
documentation before submitting a Pull Request!
Thank you for contributing to Rook! -->



**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

leseb reviewed (commented):
FYI I've made changes in the validation script to only validate certain components in https://github.com/rook/rook/pull/6474

You should probably wait a bit to get it merged and use it

leseb commented on .github/workflows/integration-test.yaml:
```diff
@@ -0,0 +1,46 @@
+on: [pull_request]
+
+jobs:
+  test:
```
```suggestion
  CephSmokeSuite:
```

subhamkrai replied:
Done.

leseb reviewed (changes requested):
Looks good so far! Can you try adding more `job`?

subhamkrai commented:
> Looks good so far! Can you try adding more `job`?

Okay, I'll try adding more `Ceph` tests.

subhamkrai commented on .github/workflows/integration-test.yaml:
```diff
@@ -0,0 +1,46 @@
+        docker images
+        docker tag $(docker images|awk '/build-/ {print $1}') rook/ceph:master
+
+    - name: TestCephSmokeSuite
```
Now, we can see logs of any pod other but for now, added only operator pod log.

leseb commented on .github/workflows/integration-test.yaml:
```diff
@@ -0,0 +1,292 @@
+      run: | 
+       cd /home/runner/work/rook/rook/tests/integration/_output/tests/
+       ls
+       FILE=$(find . ! -name '*_previous_*' -name 'TestCephMultiClusterDeploySuite_localhost_mrc-n1-system_rook-ceph-operator-*.log')
```
We should probably write a collector script at some point.

subhamkrai replied:
Agreed. I'll do it once all the tests will pass.

BlaineEXE commented:
Should we, with this PR, also set it so that the test added to GH actions don't run in Jenkins by default?

BlaineEXE commented:
> > Should we, with this PR, also set it so that the test added to GH actions don't run in Jenkins by default?
> 
> Yes, there was a separate discussion that we want to skip the Jenkins integration tests if `[test ceph]` (or `ceph:` in the PR title) and **not** `[test full]`

Sounds like that's for a different PR.

subhamkrai commented:
@travisn I have added logs of all the files.
And about snapshotter version, if we see [here](https://github.com/rook/rook/blob/master/tests/framework/utils/snapshot.go#L28) we use `"v2.1.0"` and it is passing in others PR. Also, changed the version to `"v3.0.0"` but still failing.

subhamkrai commented:
> For the mon not starting, looks like it's not finding a local volume. Is the test calling `localPathPV.sh`? Looks like that script sets up the local PV.
> 
> ```
> 27s         Warning   FailedMount            pod/rook-ceph-mon-a-5446448f78-hn7qp                                              
> MountVolume.NewMounter initialization failed for volume "local-vol1" : path "/var/lib/rook/rook-integration-test/mon1" does not exist
> 27s         Warning   FailedMount            pod/rook-ceph-mon-a-5446448f78-hn7qp                                              
> Unable to attach or mount volumes: unmounted volumes=[ceph-daemon-data], unattached volumes=[rook-ceph-mons-keyring rook-ceph-log rook-ceph-crash ceph-daemon-data default-token-gcm7k rook-config-override]: 
> timed out waiting for the condition
> ```

Yes. it is calling that script.

sp98 commented:
> The 3.0.0 snapshotter doesn't exist at that same URL, if @Madhu-1 could take a look at that test.
> 
> For the mon not starting, looks like it's not finding a local volume. Is the test calling `localPathPV.sh`? Looks like that script sets up the local PV.
> 
> ```
> 27s         Warning   FailedMount            pod/rook-ceph-mon-a-5446448f78-hn7qp                                              
> MountVolume.NewMounter initialization failed for volume "local-vol1" : path "/var/lib/rook/rook-integration-test/mon1" does not exist
> 27s         Warning   FailedMount            pod/rook-ceph-mon-a-5446448f78-hn7qp                                              
> Unable to attach or mount volumes: unmounted volumes=[ceph-daemon-data], unattached volumes=[rook-ceph-mons-keyring rook-ceph-log rook-ceph-crash ceph-daemon-data default-token-gcm7k rook-config-override]: 
> timed out waiting for the condition
> ```

Could be an issue with the [minikube](https://github.com/kubernetes/minikube/issues/4634). It wants the path to be created inside the minikube vm.  Workaround of [creating  the directory](https://github.com/kubernetes/minikube/issues/4634#issuecomment-523070730) - `/var/lib/rook/rook-integration-test/mon1  .. /mon2 , ../mon3` in minikube  fixes the issue with mons.

subhamkrai commented:
> > For the mon not starting, looks like it's not finding a local volume. Is the test calling `localPathPV.sh`? Looks like that script sets up the local PV.
> > ```
> > 27s         Warning   FailedMount            pod/rook-ceph-mon-a-5446448f78-hn7qp                                              
> > MountVolume.NewMounter initialization failed for volume "local-vol1" : path "/var/lib/rook/rook-integration-test/mon1" does not exist
> > 27s         Warning   FailedMount            pod/rook-ceph-mon-a-5446448f78-hn7qp                                              
> > Unable to attach or mount volumes: unmounted volumes=[ceph-daemon-data], unattached volumes=[rook-ceph-mons-keyring rook-ceph-log rook-ceph-crash ceph-daemon-data default-token-gcm7k rook-config-override]: 
> > timed out waiting for the condition
> > ```
> 
> Could be an issue with the [minikube](https://github.com/kubernetes/minikube/issues/4634). It wants the path to be created inside the minikube vm. Workaround of [creating the directory](https://github.com/kubernetes/minikube/issues/4634#issuecomment-523070730) - `/var/lib/rook/rook-integration-test/mon1 .. /mon2 , ../mon3` in minikube fixes the issue with mons.

@sp98 Thanks for taking a look. It seems to work now.

leseb commented on .github/workflows/integration-test.yaml:
```diff
@@ -0,0 +1,286 @@
+      run: |
+        kubectl cluster-info
+        kubectl get pods -n kube-system
+        minikube ssh -- sudo mkdir -p /var/lib/rook/rook-integration-test/mon1 /var/lib/rook/rook-integration-test/mon2 /var/lib/rook/rook-integration-test/mon3
```
You don't need that if you use the `bare` provider. Why do you use `docker` all of the sudden?

subhamkrai replied:
Yeah, I'm using `docker` only for `TestCephMultiClusterDeploySuite ` this test was failing earlier with some reason...I don't remember the error but while running the script `localPathPV.sh`. I'll try again with `bare` and see if it's working.

leseb replied:
You might as well run the `localPathPV.sh` before it helps. bare should work as all the other canary scenario use it successfully.

subhamkrai commented:
> The 3.0.0 snapshotter doesn't exist at that same URL, if @Madhu-1 could take a look at that test.
#6642

leseb commented:
> > The 3.0.0 snapshotter doesn't exist at that same URL, if @Madhu-1 could take a look at that test.
> > #6642

Ready for a rebase!

leseb commented on .github/workflows/integration-test.yaml:
```diff
@@ -0,0 +1,309 @@
+      uses: actions/upload-artifact@v2
+      if: failure()
+      with:
+        name: my-artifact
```
```suggestion
        name: ceph-multi-cluster-deploy-suite-artifact
```

subhamkrai replied:
Done.

leseb reviewed (approved):
Apply the suggestion to all the jobs

leseb commented on .github/workflows/integration-test.yaml:
```diff
@@ -240,6 +244,8 @@ jobs:
        tests/scripts/minikube.sh helm
        tests/scripts/helm.sh up
+       sed -i "s|deviceFilter:  ''|deviceFilter: $(lsblk|awk '/14G/ {print $1}'| head -1)|g" tests/framework/installer/ceph_manifests.go
+       sed -i "s|for i := 0; i < 30; i++ {|for i := 0; i < 60; i++ {|g" tests/framework/utils/k8s_helper.go
```
why not just edit `tests/framework/utils/k8s_helper.go` directly?

subhamkrai replied:
Done.

subhamkrai commented:
> Next let's consider adding the CephMgrSuite, but keep it optional since it runs against ceph master.

#6669
